File: docs/garden/src/haskell-purity-laziness/lazy_gen.py
# Python: строгие аргументы, лень — только явная, через генераторы.
from itertools import count, islice


def trace(name, value):
    print(f"  вычисляю {name}")
    return value


def const_one(_):
    return 1


# Аргумент вычисляется до вызова, даже если функции он не нужен.
print("const_one:", const_one(trace("аргумент", 42)))


def boom():
    raise RuntimeError("аргумент вычислен")


try:
    const_one(boom())
except RuntimeError as e:
    print("строгость:", e)


# Генератор — явно отложенная последовательность.
def fibs():
    a, b = 0, 1
    while True:
        yield a
        a, b = b, a + b


def primes():
    found = []
    for n in count(2):
        if all(n % p for p in found):
            found.append(n)
            yield n


print(list(islice(fibs(), 10)))
print(list(islice(primes(), 10)))

# Список вычисляется целиком сразу, генераторное выражение — по требованию.
squares_list = [trace(f"{x}^2", x * x) for x in range(3)]
print("список построен")
squares_gen = (trace(f"{x}^2", x * x) for x in range(3))
print("генератор создан")
print(next(squares_gen))

# Генератор одноразовый: в отличие от ленивого списка Haskell, он не мемоизирует.
g = (x for x in range(3))
print(list(g), list(g))

File: docs/garden/src/haskell-purity-laziness/lazy_gen.js
// JavaScript: аргументы вычисляются до вызова; лень — через генераторы
// или явные «thunk» — функции без параметров.
function trace(name, value) {
  console.log(`  вычисляю ${name}`);
  return value;
}
const constOne = (_x) => 1;
console.log("constOne:", constOne(trace("аргумент", 42)));

// Отложенное значение вручную: функция без параметров + мемоизация.
function lazy(f) {
  let done = false, value;
  return () => {
    if (!done) { value = f(); done = true; }
    return value;
  };
}
const expensive = lazy(() => trace("expensive", 6 * 7));
console.log("thunk создан");
console.log(expensive(), expensive()); // вычисляется один раз

function* fibs() {
  let [a, b] = [0n, 1n];
  for (;;) { yield a; [a, b] = [b, a + b]; }
}
function take(n, it) {
  const out = [];
  for (const x of it) { if (out.length >= n) break; out.push(x); }
  return out;
}
console.log(take(10, fibs()).join(" "));

// && и || — единственные «нестрогие» места: правый операнд может не вычисляться.
false && trace("правый операнд", true);
console.log("после false && ...");
